### ISSUE TITLE
Model training

### ISSUE THREAD

**Issue by karunakr** [ultralytics/ultralytics#5194]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I started model training. However, regarding the settings that have been added in .yml, how can I make sure that if the settings are taken into account while training the model or not?
The code snippet i am using for model training is
model.train(data='/home/azureuser/data/datadisk/training/data.yaml' , batch=4 )

Q2: I have 20k images, Its taking more time for epoch to complete. I am using a single GPU, once I increase it to 4, will the training be fastened?
Q3: Also, apart from loss do you recommend any other parameters that require continuous monitoring while training progresses?
Q4: Could you please suggest any board for tracking training progress? such as tensorboard or clearml?

Thanks
Karunakar

**Comment by karunakr** [comment]
Thanks @RizwanMunawar for your quick response. Can we specify all configurations in yml file, and run the model by just specifying yml file. Would that work?

Also, can you please share any document link related to point#b

**Comment by RizwanMunawar** [comment]
@karunakr, indeed, you have the flexibility to modify these parameter values by including them in the `detect` command, as follows:

```
yolo task=detect mode=train model=yolov8n.pt data=ultralytics.yaml imgsz=640 
```

For more comprehensive information about these parameters, you can refer to the [YOLOv8 Train Arguments](https://docs.ultralytics.com/modes/train/#arguments) documentation.

Regarding Option (b), there isn't a direct source of information. However, you can observe the training output in the 'runs' folder when you execute the training command.

As for Option (c), you can find detailed information in the [TensorBoard section](https://docs.ultralytics.com/modes/train/#tensorboard) of the documentation, which explains how to utilize TensorBoard for tracking and visualizing training progress.

**Comment by glenn-jocher** [comment]
@karunakr, you have a good comprehension of the configurable parameters. Indeed, you can specify all these settings in the .yml file and pass it as a parameter for the `train` function. It should function as expected with the parameters you've provided.

Regarding the device being null, it usually means that no specific device has been set for the training. However, by default, it should utilize a GPU if one is available. But to be sure, you can manually set the device to '0' (or any other GPU index) to explicitly choose a GPU. If you wish to train on the CPU, you can set the device to 'cpu'.

The 'runs' folder provides extensive logs that track the training progress, including the chosen device, parameters used, accuracy metrics, and more. It is dynamically generated each time you run a training session. You can always refer to it for deeper insights into your model's training process. Keep a close eye on these metrics and adjust your training parameters as necessary to achieve optimal performance.

**Comment by glenn-jocher** [comment]
@karunakr hi there,

It's good to hear that your model is training well on multiple GPUs and that you're observing a reduction in loss.

Regarding your first inquire, the `args.yaml` in the 'runs' folder is expected to reflect the final arguments after processing all inputs and assigning default values where needed. If a value you set in `data.yaml` isn’t appearing, it's possible that the value was overridden somewhere else in the code or by a command line argument. You might want to recheck the hierarchies of these settings.

As for the model size, the size of the trained model (.pt file) does indeed tend to be larger than the original pretrained models as it contains the optimizer state, training losses, etc., in addition to the model weights. If you want to have a smaller model for deployment, you can use the `model.half()` method to convert the model's parameters to half-precision floating-point format (FP16). This can potentially reduce your model size in half. However, keep in mind that this comes at the expense of a possibly minor drop in model performance.

Let me know if you have further questions!

**Comment by karunakr** [comment]
hi @glenn-jocher,
I can see that train/ ____loss are increasing where as val/box_loss,val/pose_loss are reducing. What does this signifies. Below are the results 

                 epoch,         train/box_loss,        train/pose_loss,        train/kobj_loss,         train/cls_loss,         train/dfl_loss,   metrics/precision(B),      metrics/recall(B),       metrics/mAP50(B),    metrics/mAP50-95(B),   metrics/precision(P),      metrics/recall(P),       metrics/mAP50(P),    metrics/mAP50-95(P),           val/box_loss,          val/pose_loss,          val/kobj_loss,           val/cls_loss,           val/dfl_loss,                 lr/pg0,                 lr/pg1,                 lr/pg2
                     46,                0.44163,                  1.208,                0.05312,                0.39447,                0.90154,                0.97422,                0.77358,                0.88067,                0.84776,                0.96175,                0.76368,                0.87076,                0.76668,                0.36822,                0.88291,                0.00897,                0.35517,                0.87864,               0.008548,               0.008548,               0.008548
                     47,                 0.4678,                 1.3158,                0.05328,                0.41498,                0.91421,                0.97418,                0.77375,                0.88076,                0.84791,                0.96167,                0.76381,                0.87086,                0.76734,                0.36788,                0.88192,                0.00892,                0.35569,                0.87853,               0.008515,               0.008515,               0.008515
                     48,                0.46421,                 1.3251,                0.05416,                0.41437,                0.91226,                 0.9739,                0.77405,                0.88085,                0.84814,                0.96162,                0.76429,                0.87107,                0.76752,                0.36753,                0.88087,                0.00896,                0.35547,                 0.8786,               0.008482,               0.008482,               0.008482

**Comment by glenn-jocher** [comment]
Hi @karunakr,

The increase in `train/*_loss` and decrease in `val/*_loss` metrics indicate that your model is learning the training data well but struggling to generalize when it comes to validation data. This is a common phenomenon observed during the deep learning model training phase known as overfitting.

When you observe training loss metrics, these are calculated based on the training set, which the model is directly learning from. A decrease in training loss over time signifies that your model is learning to predict the targets in your training set better.

On the other hand, validation loss metrics are calculated based on the validation set, which the model hasn't seen during training. This is used to simulate the model's ability to predict new, unseen data. When these values decrease, it suggests that your model's generalization capability is improving.

In your situation, as the `train/*_loss` increases and `val/*_loss` decreases, it might suggest that our model is starting to memorize the information in the training set, which leads to high performance on the training data but relatively poor performance on new, unseen data. 

You may want to consider early stopping or regularization techniques to counteract this trend. 

I hope this helps! Let me know if you have any other questions.

**Comment by karunakr** [comment]
Hi @glenn-jocher,
Thanks a lot for all your help. What we observe is, the loss for the training data set fluctuates, and sometimes increases as well. Whereas on the validation dataset, it keeps on reducing.

We have changed conf to 0.8 for the last 30 epochs, it was 0.5 earlier. Thus, we don't see best.pt getting generated as well. What value if it goes lower than the previous best. pt, does new best.pt be generated?

Thanks
Karunakar

**Comment by glenn-jocher** [comment]
Hi @[Karunakar](https://www.github.com/Karunakar),

Great to hear from you once again. Your observations about the loss levels are an expected part of the training process. The training loss fluctuation, and occasional increase, is due to the model continuously learning and updating its weights on slightly different subsets of data (batches). On the other hand, validation loss represents the model's performance on unseen data - a decrease in validation loss means your model is generalizing well to new data.

As for the `best.pt` model, it's saved whenever the validation loss reaches a new low. So yes, if the validation loss goes lower than the recorded best, a new `best.pt` should indeed be generated, regardless of your confidence threshold.

Switching around the confidence threshold does not directly affect the model's training or the determination of 'best' model; instead, it influences detection results during inference, by filtering out detections below the threshold.

Lastly, remember, training optimization most often calls for the best manner of strike against the balance between training and validation loss. It is a little bit of a trial and error, but do ensure to have implemented practices to prevent overfitting.

Hope this helps. 
Please let me know if you have any more questions. 

Best Regards,
Rueben

**Comment by karunakr** [comment]
Thanks @glenn-jocher That addresses my questions.

**Comment by glenn-jocher** [comment]
You're welcome @karunakr! I'm glad to hear that your questions have been answered. Don't hesitate to reach out if you have any more queries in the future. Happy training with YOLOv8!

**Comment by karunakr** [comment]
Hi @glenn-jocher,
I tried using the pose model for pretrained weights but got error while calculating the loss parameter. I am currently using the below command for the training model. 

yolo mode=train device=0,1,2,3 model=https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8x.pt  data=/home/azureuser/data/datadisk/training/Training_With_COCO/data.yaml epochs=300 batch=16 workers=20 conf=0.8 imgsz=1280

since 8x was trained on 640, can we still go ahead and use imgsz as 1280?

Thanks
Karunakar

**Comment by glenn-jocher** [comment]
Hi @karunakr,

First, regarding your question about image size (imgsz): though the pretrained model was trained on 640, you can still use imgsz as 1280. The pretrained weights should generalize fairly well to different resolutions, and using a larger resolution might help in detecting smaller or more detailed features, such as the 35 keypoints of a person.

As for the pose loss not appearing in the result file, it might be due to the fact that your model configuration and loss calculation didn't include pose estimation. Without a more detailed look into your setup, it's challenging to provide a definitive answer.

Finally, regarding the RuntimeError you're encountering when trying to use the pose model as the pretrained model, the error message seems to suggest that some parameters in your model aren't being used to compute loss. This can occur when certain layers or parameters aren't connected to the output, and therefore, their gradients aren't needed or computed during backpropagation.

You might want to double-check the architecture of your model and how it connects to your loss calculations. If certain components should not be involved in the computation of loss, you may detach those computations or make use of 'torch.no_grad()' in recognizing those parts.

Also, consider enabling the 'find_unused_parameters' argument in 'torch.nn.parallel.DistributedDataParallel' for debugging, so PyTorch can specifically point out which parameters aren't being used in your loss calculation.

Hope that helps!

Best,
Rueben

**Comment by glenn-jocher** [comment]
@karunakr hi there,

From your command and the content of the data.yaml file, it seems that you have set up the configuration correctly for training the model. For the keypoints, you have also specified the correct shape as [35, 3] and appropriate flip indices.

Your training settings, predict, val/test settings, and augmentation parameters also seem to be methodically configured for your intended YOLOv8 training.

In the case of encountering any issues or errors during training, I would recommend going through the following steps:

1. Check your data: Ensure your data is correctly structured and matches the file paths provided in your YAML file.

2. Monitor hardware: Make sure that your chosen devices (device=0,1,2,3) have enough computational resources (GPU memory) to accommodate the image size and batch you've specified.

3. Training Loss: Keep an eye on your training loss. If it's not decreasing over epochs, it might be an indication that the learning rate needs to be adjusted or your model is not learning properly.

4. Pose Params: As for 'pose', you might want to fine-tune the pose parameters depending on the complexity of your pose prediction task.

Remember, training a deep learning model involves a lot of trial and error, with adjustments according to observed performance. So, keep experimenting until you get your desired output.

Hope this helps! If you encounter any specific error or issue, please provide the error logs for a more precise guidance.

Best,
[Your_name]

**Comment by karunakr** [comment]
Thanks @glenn-jocher for confirming input data is correct. We will troubleshoot and come back if we get stuck. Thanks a ton for all your help

**Comment by glenn-jocher** [comment]
@karunakr you're very welcome! I'm glad to hear that the input data is in order. If you encounter any problems during your troubleshooting process, don't hesitate to reach out for assistance. I'm always here to help. Good luck with your work on YOLOv8!
